DTensor: when forcing tangents to be correct, force syncing collectives at graph boundaries

bdhirsh opened the pull request:
This is enough to fix https://github.com/pytorch/pytorch/issues/118664.

The problem is that when we "correct" tangents to have a non-partial placement at trace-time, we do so by calling `redistribute()` - which can return an `AsyncCollectiveTensor`.

This means that we now have a tangent at trace-time that looks like `DTensor(AsyncCollectiveTensor(FakeTensor))`.

This fails for a bunch of reasons. But I think the more significant reason is that support for `AsyncCollectiveTensors` as graph boundaries is not great today:

(1) I think getting async outputs of compile working fully will be a non-trivially amount of work

(2) forcing the sync here ensures correctness. And while it has a performance cost, most of this performance can be clawed back by removing graph breaks.

I fixed this in the current PR by providing a way to force `redistribute()` to sync its output - so the result of `DTensor.restribute()` is guaranteed not to have an `AsyncCollectiveTensor`.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119948
* #118802
* __->__ #118671
* #118670
* #118669
* #119947
* #118803



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

bdhirsh committed:
DTensor: when forcing tangents to be correct, for syncing collectives at graph boundaries

[ghstack-poisoned]

wconstab commented on torch/distributed/_tensor/redistribute.py:
```diff
@@ -261,6 +262,8 @@ def forward(  # type: ignore[override]
         local_tensor = input._local_tensor
         output = redistribute_local_tensor(local_tensor, current_spec, target_spec)
+        if force_wait:
+            output = torch.distributed._functional_collectives.wait_tensor(output)
```
is it the case that output may or may not be an ACT, so we can't just call output.wait?  (iirc we added some api to ACT for this purpose)

bdhirsh replied:
Hmm I just wasn’t sure if wait() would properly no-op if you pass it a plain tensor. But if that should work, I’ll try it and see if anything fails

wanchaol replied:
I think this is technically a no-op as `wait_tensor` would check if there's a work associate with the tensor and if there is then call `wait()`. 

But I think it would still be good to have a explicitly check `if isinstance(output, AsyncCollectiveTensor)` then call wait.

wconstab reviewed (commented):
hm, approach sounds good to me. but seems like tests are failing, maybe re-review after fixed?

bdhirsh commented:
Yep- there are a bunch of CI failures throughout this stack- I’ll start digging into them this week and give a shout when CI is passing / if there are any major problems

bdhirsh committed:
Update on "DTensor: when forcing tangents to be correct, force syncing collectives at graph boundaries"

This is enough to fix https://github.com/pytorch/pytorch/issues/118664.

The problem is that when we "correct" tangents to have a non-partial placement at trace-time, we do so by calling `redistribute()` - which can return an `AsyncCollectiveTensor`.

This means that we now have a tangent at trace-time that looks like `DTensor(AsyncCollectiveTensor(FakeTensor))`.

This fails for a bunch of reasons. But I think the more significant reason is that support for `AsyncCollectiveTensors` as graph boundaries is not great today:

(1) I think getting async outputs of compile working fully will be a non-trivially amount of work

(2) forcing the sync here ensures correctness. And while it has a performance cost, most of this performance can be clawed back by removing graph breaks.

I fixed this in the current PR by providing a way to force `redistribute()` to sync its output - so the result of `DTensor.restribute()` is guaranteed not to have an `AsyncCollectiveTensor`.





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

bdhirsh committed:
Update on "DTensor: when forcing tangents to be correct, force syncing collectives at graph boundaries"

This is enough to fix https://github.com/pytorch/pytorch/issues/118664.

The problem is that when we "correct" tangents to have a non-partial placement at trace-time, we do so by calling `redistribute()` - which can return an `AsyncCollectiveTensor`.

This means that we now have a tangent at trace-time that looks like `DTensor(AsyncCollectiveTensor(FakeTensor))`.

This fails for a bunch of reasons. But I think the more significant reason is that support for `AsyncCollectiveTensors` as graph boundaries is not great today:

(1) I think getting async outputs of compile working fully will be a non-trivially amount of work

(2) forcing the sync here ensures correctness. And while it has a performance cost, most of this performance can be clawed back by removing graph breaks.

I fixed this in the current PR by providing a way to force `redistribute()` to sync its output - so the result of `DTensor.restribute()` is guaranteed not to have an `AsyncCollectiveTensor`.





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

bdhirsh committed:
Update on "DTensor: when forcing tangents to be correct, force syncing collectives at graph boundaries"

This is enough to fix https://github.com/pytorch/pytorch/issues/118664.

The problem is that when we "correct" tangents to have a non-partial placement at trace-time, we do so by calling `redistribute()` - which can return an `AsyncCollectiveTensor`.

This means that we now have a tangent at trace-time that looks like `DTensor(AsyncCollectiveTensor(FakeTensor))`.

This fails for a bunch of reasons. But I think the more significant reason is that support for `AsyncCollectiveTensors` as graph boundaries is not great today:

(1) I think getting async outputs of compile working fully will be a non-trivially amount of work

(2) forcing the sync here ensures correctness. And while it has a performance cost, most of this performance can be clawed back by removing graph breaks.

I fixed this in the current PR by providing a way to force `redistribute()` to sync its output - so the result of `DTensor.restribute()` is guaranteed not to have an `AsyncCollectiveTensor`.





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

bdhirsh committed:
Update on "DTensor: when forcing tangents to be correct, force syncing collectives at graph boundaries"

This is enough to fix https://github.com/pytorch/pytorch/issues/118664.

The problem is that when we "correct" tangents to have a non-partial placement at trace-time, we do so by calling `redistribute()` - which can return an `AsyncCollectiveTensor`.

This means that we now have a tangent at trace-time that looks like `DTensor(AsyncCollectiveTensor(FakeTensor))`.

This fails for a bunch of reasons. But I think the more significant reason is that support for `AsyncCollectiveTensors` as graph boundaries is not great today:

(1) I think getting async outputs of compile working fully will be a non-trivially amount of work

(2) forcing the sync here ensures correctness. And while it has a performance cost, most of this performance can be clawed back by removing graph breaks.

I fixed this in the current PR by providing a way to force `redistribute()` to sync its output - so the result of `DTensor.restribute()` is guaranteed not to have an `AsyncCollectiveTensor`.





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

bdhirsh committed:
Update on "DTensor: when forcing tangents to be correct, force syncing collectives at graph boundaries"

This is enough to fix https://github.com/pytorch/pytorch/issues/118664.

The problem is that when we "correct" tangents to have a non-partial placement at trace-time, we do so by calling `redistribute()` - which can return an `AsyncCollectiveTensor`.

This means that we now have a tangent at trace-time that looks like `DTensor(AsyncCollectiveTensor(FakeTensor))`.

This fails for a bunch of reasons. But I think the more significant reason is that support for `AsyncCollectiveTensors` as graph boundaries is not great today:

(1) I think getting async outputs of compile working fully will be a non-trivially amount of work

(2) forcing the sync here ensures correctness. And while it has a performance cost, most of this performance can be clawed back by removing graph breaks.

I fixed this in the current PR by providing a way to force `redistribute()` to sync its output - so the result of `DTensor.restribute()` is guaranteed not to have an `AsyncCollectiveTensor`.





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/redistribute.py:
```diff
@@ -252,6 +252,7 @@ def forward(  # type: ignore[override]
         input: "dtensor.DTensor",
         device_mesh: DeviceMesh,
         placements: Tuple[Placement, ...],
+        force_wait: bool = False,
```
I think let's call this `async_output` and make it by default == True?

wanchaol reviewed (approved):
sounds good to me! some comments inlined.

wanchaol commented:
Also seems like the test failures are real?

bdhirsh committed:
Update on "DTensor: when forcing tangents to be correct, force syncing collectives at graph boundaries"

This is enough to fix https://github.com/pytorch/pytorch/issues/118664.

The problem is that when we "correct" tangents to have a non-partial placement at trace-time, we do so by calling `redistribute()` - which can return an `AsyncCollectiveTensor`.

This means that we now have a tangent at trace-time that looks like `DTensor(AsyncCollectiveTensor(FakeTensor))`.

This fails for a bunch of reasons. But I think the more significant reason is that support for `AsyncCollectiveTensors` as graph boundaries is not great today:

(1) I think getting async outputs of compile working fully will be a non-trivially amount of work

(2) forcing the sync here ensures correctness. And while it has a performance cost, most of this performance can be clawed back by removing graph breaks.

I fixed this in the current PR by providing a way to force `redistribute()` to sync its output - so the result of `DTensor.restribute()` is guaranteed not to have an `AsyncCollectiveTensor`.





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
hmm there was one test failure because the test that I added in the previous PR wasn't actually fixed until this PR. But.. I think that was the only one. I'll let CI finish and see if there's more fallout.

bdhirsh committed:
Update on "DTensor: when forcing tangents to be correct, force syncing collectives at graph boundaries"

This is enough to fix https://github.com/pytorch/pytorch/issues/118664.

The problem is that when we "correct" tangents to have a non-partial placement at trace-time, we do so by calling `redistribute()` - which can return an `AsyncCollectiveTensor`.

This means that we now have a tangent at trace-time that looks like `DTensor(AsyncCollectiveTensor(FakeTensor))`.

This fails for a bunch of reasons. But I think the more significant reason is that support for `AsyncCollectiveTensors` as graph boundaries is not great today:

(1) I think getting async outputs of compile working fully will be a non-trivially amount of work

(2) forcing the sync here ensures correctness. And while it has a performance cost, most of this performance can be clawed back by removing graph breaks.

I fixed this in the current PR by providing a way to force `redistribute()` to sync its output - so the result of `DTensor.restribute()` is guaranteed not to have an `AsyncCollectiveTensor`.





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

bdhirsh committed:
Update on "DTensor: when forcing tangents to be correct, force syncing collectives at graph boundaries"

This is enough to fix https://github.com/pytorch/pytorch/issues/118664.

The problem is that when we "correct" tangents to have a non-partial placement at trace-time, we do so by calling `redistribute()` - which can return an `AsyncCollectiveTensor`.

This means that we now have a tangent at trace-time that looks like `DTensor(AsyncCollectiveTensor(FakeTensor))`.

This fails for a bunch of reasons. But I think the more significant reason is that support for `AsyncCollectiveTensors` as graph boundaries is not great today:

(1) I think getting async outputs of compile working fully will be a non-trivially amount of work

(2) forcing the sync here ensures correctness. And while it has a performance cost, most of this performance can be clawed back by removing graph breaks.

I fixed this in the current PR by providing a way to force `redistribute()` to sync its output - so the result of `DTensor.restribute()` is guaranteed not to have an `AsyncCollectiveTensor`.





cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]
